Add i18n

asedmammad opened the pull request:
## Pull Request Checklist

- [x] **Description:** Briefly describe the changes in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?

---

## Description

* This pull request introduces internationalization (i18n) to the project, including the addition of a setting input to determine the UI language. When a language is selected, the user interface adjusts accordingly. Internationalization is done using `i18next`, and could be used in svelte files like this:
```javascript
// const i18n = getContext('i18n');
$i18n.t('translation key')
```
* This needs `translation key` to be present in translation files located at `src/lib/i18n/locales` for each language. Translation are specified in JSON format. Adding new `translation key`s to these translation files can be done either manually or by issuing the following command in the root directory of the project:
```
npm run i18n:parse
```

* This scans `js`/`svelte` files in the `src` directory for translation keys and adds missing keys for available languages.  Available languages are specified in `src/lib/i18n/locales/languages.json`. 

* Also we utilize `i18next-browser-languagedetector` to detect current user's language and set it in a svelte store which will update the UI with the detected language. Detection is done in this order: 
1. `querystring` (e.g `lang=en`)
2. `localStorage`
3. `navigator`

* We use current string literals as translation keys, so that the keys themselves can act as fallback translation strings and also help in the process of adding translation values.
---

### Changelog Entry

### Added

- Added https://github.com/i18next/i18next library dependency for i18n
- Added https://github.com/i18next/i18next-browser-languageDetector, i18next plugin to detect user's language
- Added https://github.com/i18next/i18next-resources-to-backend, i18next plugin to use json resource files as translation provider
- Added https://github.com/i18next/i18next-parser as dev dependency to automatically insert missing keys into json translation files.
- Initial setup for i18n
- Added Farsi (Persian) translation
- Added Deutsch (German ) translation (by @jannikstdl)
- Added Ukranian translation (by @SimonOriginal )
- Added French translation (by @BYTOOX )
- Added Simplified Chinese translation (by @kawais)

jannikstdl commented:
I initialized german in my clone of your fork. 
We can work together in 
I can not push into this PR so maybe the best way is that you give me access to your fork and i push my changes to yours.

jannikstdl commented on src/lib/translations/fa/app.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"New Chat": "چت جدید",
 	"Modelfiles": "فایل‌های مدل",
-	"Placeholder Greeting": "كيف يمكنني مساعدتك اليوم؟",
```
Sorry this was DeepL Translator :D

justinh-rahb commented:
I don't mind pulling this PR and building to test, let me know when it's ready 👍

jannikstdl commented:
@asedmammad i would do the german translation and locale variable definition of the settings

tjbck commented on package.json:
```diff
@@ -50,7 +50,8 @@
 		"katex": "^0.16.9",
 		"marked": "^9.1.0",
 		"svelte-french-toast": "^1.2.0",
+		"sveltekit-i18n": "^2.4.2",
```
Thanks for the PR! `sveltekit-i18n` seems like it's no longer maintained, is there a way to use framework agnostic `i18n` library?

asedmammad replied:
`iI18next` seems like a good suggestion.

We can explore `i18next` further or research other framework agnostic solutions, @tjbck  Do you have any specific solution in mind?

tjbck replied:
`i18next` looks good to me as well!

asedmammad replied:
I will investigate `i18next` and attempt to create an initial configuration for it.

asedmammad replied:
@tjbck Basic i18next setup is done, I hope adding `i18next-browser-languagedetector` and `i18next-resources-to-backend` is ok. Please let me know if these additions are not acceptable or if there's a preferred alternative approach that should be taken instead. 
Next step can be scanning the whole project for string literals to be added to json locale files.

tjbck replied:
@asedmammad Thanks for all the effort! They both seem promising but are there any alternatives that I should be aware of?

asedmammad replied:
@tjbck My pleasure!
Manually detecting the active language using localStorage, query param, and `window.navigator.language` is an alternative approach.
Other i18next language detector plugins can be found here: https://www.i18next.com/overview/plugins-and-utils#language-detector

Having translations in project json files seems a reasonably good option, though adds a few KBs to the build. If the bundle size is/becomes a concern, we could split the chunks to avoid loading all translation files at first load.  I will compare the bundle size against the latest `dev` commit and update here.
Another option is to load translations over http from the project's backend. 
i18next provides some translation backend providers here: 
https://www.i18next.com/overview/plugins-and-utils#backends

BYTOOX commented:
Hey, this project is really interesting. If you need some french translator, i'm up to the task.

jannikstdl commented:
> SSR seems to be inactive (`src/routes/+layout.js:10`), Isn't it? Initial locale value in the setting can be set to `navigator.language || 'en' `.

@asedmammad Somehow this was not working with the old framework. But now with i18next it seems to be working! It recognized my locale.

Status Update:
- All text should be migrated to i18next variables
- Python language dependend texts like the RAG Template, Prompt Suggestions and Error Messages are not

@tjbck i changed the theme svgs to unicode icons because with larger text it was off (as far as i know you cant add flexible margin depending on the textlength of html selects):
![image](https://github.com/open-webui/open-webui/assets/69747628/dc6e77e5-a67f-4dd0-9f77-70fac97426bb)
to
![image](https://github.com/open-webui/open-webui/assets/69747628/b2c34150-cd77-4cbf-be80-a0478b0df30c)

Also on Windows (Maybe Linux too) the text on selects with darkmode is not readable:
![image](https://github.com/open-webui/open-webui/assets/69747628/59398f19-1392-4751-9e00-9296e05717a6)

I would fix this adding dark:bg-gray-900 to the classes, if that doesent affect the bg-transparent (which eventually only works on macos with selects, or maybe macos selects are transparent by default?)

jannikstdl commented:
@justinh-rahb you could test this now!

SimonOriginal commented:
I made a translation file based on English localization into Ukrainian. What do I need for you to include it in your branch?

SimonOriginal commented:
I'll give you the file via Google Drive. Ok? - https://drive.google.com/drive/folders/17Yx3d0b4JqHr9va7kvHkQDzhbcrHQS3V?usp=sharing

jannikstdl commented:
@SimonOriginal Thanks, added :)

jannikstdl commented:
@BYTOOX we could add french directly too

justinh-rahb commented:
Now, if only I spoke German 🤔 Looking good to me everyone, except the FR translation doesn't appear to do anything?

![Screenshot 2024-03-05 at 10 00 40 AM](https://github.com/open-webui/open-webui/assets/52832301/174cbde1-c2a9-4a28-a46c-808bc350c9f9)

jannikstdl commented:
@justinh-rahb yes that's normal because we don't have a French translated file yet. 

Something I noticed:
- If you change the language the title generation prompt only changes if you go to the setting under "interface" and delete the contend - hit save and then go back to the interface tab. 
Only then it gets changed to the selected language.  

- For future translators because of zu updates it gets hard to creep track of which textelements can be translated. Maybe we can automate that when we add a i18n variable in the code it automatically gets added to the common.json fils.  I think the plugin "i18n parser" for i18next does that.

BYTOOX commented:
> @BYTOOX we could add french directly too

Yup, didn't follow the repository, damn you guys are working fast. My translation is not yet ready, will finish it asap.

asedmammad commented:
> * For future translators because of updates it gets hard to keep track of which textelements can be translated. Maybe we can automate that whenever we add a i18n variable in the code it automatically gets added to the common.json files.  I think the plugin "i18n parser" for i18next does that.

* Yes, will explore `i18n parser` to hopefully solve that.


* Another issue is that currently a list of languages is hardcoded in the settings. We need to find a way to populate them dynamically based on the available translations.
* Also I was thinking about maybe renaming the json files from `common.json` to maybe something like `base.json`, `webui.json` or some other name. Common implies that there is/will be other namespaces that are specific to other parts of the project.
* And typing `i18n.t` feels like a pain! `locale.t` seems easier to me! What do you think? @jannikstdl

BYTOOX commented:
Here is the french translation file. https://www.swisstransfer.com/d/8129ba9d-66d3-494a-8d33-bb7ee6f423fe

asedmammad commented:
> Here is the french translation file. https://www.swisstransfer.com/d/8129ba9d-66d3-494a-8d33-bb7ee6f423fe

@BYTOOX  Thanks, added

jannikstdl commented:
@asedmammad i think there is no native way to get all the languages which have a translation json? 

At the moment I don't have a idea how to manage that dynamically exept scanning the directory in a prebuilt script. 
Any ideas?

asedmammad commented:
> @asedmammad i think there is no native way to get all the languages which have a translation json?
> 
> At the moment I don't have a idea how to manage that dynamically exept scanning the directory in a prebuilt script. Any ideas?

No, not really. But on the other hand we need a way to map ISO language codes to the language titles. So I add  another json file which provides available languages with respective titles for now.

jannikstdl commented:
@asedmammad yes, that looks like a good compromise, i think that should be easy to understand for future translators. Great work!

asedmammad commented:
Added `i18next-parser` to the PR. This makes it easier to update json translation files by automatically extracting keys from `js` and `svelte` files and adding them to all available translation files.
Currently this can be done using the following command in the root directory of the project:
```
npx i18next --config i18next-parser.config.ts
```

SimonOriginal commented:
I've updated the Ukrainian translation file. https://drive.google.com/file/d/1x45vw5UUl_jreZ7MhhYfWa6GBzbn9vv-/view?usp=sharing

jannikstdl commented:
@asedmammad added a contribution info in the settings under the language selector.

justinh-rahb commented:
FYI, Docker build is failing at the present moment:

![Screenshot 2024-03-07 at 11 13 07 AM](https://github.com/open-webui/open-webui/assets/52832301/7ffd0772-70e4-4f1d-9c2b-d303000be1d7)

SimonOriginal commented:
When analyzing the Ukrainian version of the translation, some fields were found to be missing. Improvements were made in the description to convey the essence more accurately and clearly. - https://drive.google.com/drive/folders/17Yx3d0b4JqHr9va7kvHkQDzhbcrHQS3V?usp=sharing

justinh-rahb commented:
French translation missing some strings:

![Screenshot 2024-03-07 at 11 27 49 AM](https://github.com/open-webui/open-webui/assets/52832301/851a3f14-88db-4a38-aab2-e2ec557793bc)

jannikstdl commented:
@asedmammad maybe we should set the Admin, Pending, User to lowercase so the translation appears in the Admin Panel too.

justinh-rahb commented:
> > EDIT: this is broken no matter the language.
> 
> @justinh-rahb This should be fixed now for English, Other languages need the translation (and should fallback to `en`) I will check for other possible empty translations.

Building as we speak, will let you know if I spot anything

jannikstdl commented:
Like this:
![image](https://github.com/open-webui/open-webui/assets/69747628/6fb37da8-419b-4618-aba6-8783740aefc3)

jannikstdl commented:
> maybe we should set the Admin, Pending, User to lowercase so the translation appears in the Admin Panel too.

Had this on my local branch but now its gone. I will fix this.

asedmammad commented:
I fixed empty translations in `en`, other translations have empty values, I should change `i18next-parser` configuration to maybe add English translation as a default value.

jannikstdl commented:
> Had this on my local branch but now its gone. I will fix this.

Okay now i know why its gone. We pushed almost at the same time and it was already in the PR with [12c74b2](https://github.com/open-webui/open-webui/pull/926/commits/12c74b2cd1d9e28f76e676779e84e33ac755cc46) 

Lost my mind for 10min, but we should be fine now :D We could add a CHANGELOG.md entry, but maybe this is for tim when he adds this PR.

asedmammad commented:
> > Had this on my local branch but now its gone. I will fix this.
> 
> Okay now i know why its gone. We pushed almost at the same time and it was already in the PR with [12c74b2](https://github.com/open-webui/open-webui/pull/926/commits/12c74b2cd1d9e28f76e676779e84e33ac755cc46)
> 
> Lost my mind for 10min, but we should be fine now :D We could add a CHANGELOG.md entry, but maybe this is for tim when he adds this PR.

Oops! :D
Changed `i18next-parser` configuration, we should be okay. I will test this further to make sure I'm not missing anything here.

tjbck commented on src/lib/i18n/locales/fr/translation.json:
```diff
@@ -0,0 +1,339 @@
+	"WebUI Settings": "Paramètres WebUI",
+	"What’s New in": "Quoi de neuf dans",
+	"When history is turned off, new chats on this browser won't appear in your history on any of your devices.": "Lorsque l'historique est désactivé, les nouveaux chats sur ce navigateur n'apparaîtront pas dans votre historique sur aucun de vos appareils.",
+	"Whisper (Local)": "Chuchotement (Local)",
```
This should probably still be "whisper" ahah

SimonOriginal commented:
I'll try to update the Ukrainian translation this evening. I saw that new keys appeared there that were not translated.

SimonOriginal commented:
I have updated the translation - https://drive.google.com/drive/u/0/folders/17Yx3d0b4JqHr9va7kvHkQDzhbcrHQS3V. 
Found several issues with applying the translation (screenshots attached):
![Image](https://github.com/open-webui/open-webui/assets/94782611/4ffffea2-30d6-4833-ae8b-196c1921106a)
![Image](https://github.com/open-webui/open-webui/assets/94782611/6f367378-3d20-4fe8-b606-01af5fe62941)
![Image](https://github.com/open-webui/open-webui/assets/94782611/fb3651f0-daa0-4f7a-8c00-d432af3a3c2e)
![Image](https://github.com/open-webui/open-webui/assets/94782611/db0f2171-99cc-49df-ada1-f0d55c121b8b)
![Image](https://github.com/open-webui/open-webui/assets/94782611/08498ea1-fc3f-46d4-b454-2833b2bd658f)
![Image](https://github.com/open-webui/open-webui/assets/94782611/7e400360-348a-431e-a221-f063dd1c26e2)
![Image](https://github.com/open-webui/open-webui/assets/94782611/62ca347c-cb8a-49af-a31f-a1cbc48d7ec5).

There is an issue with font consistency (example of the issue screenshot):
![Image](https://github.com/open-webui/open-webui/assets/94782611/6094ee09-3f36-44b7-ad11-8df5f3f754fb)

justinh-rahb commented on src/lib/components/chat/Settings/Images.svelte:
```diff
@@ -136,37 +138,39 @@
-						<option value="">Default (Automatic1111)</option>
-						<option value="openai">Open AI (Dall-E)</option>
+						<option value="">{$i18n.t('Default (Automatic1111)')}</option>
+						<option value="openai">{$i18n.t('Open AI (Dall-E)'}</option>
```
There's an improper closing bracket, this should be:
```html
<option value="openai">{$i18n.t('Open AI (Dall-E)')}</option>
```

tjbck commented:
Ready to be merged?

SimonOriginal commented:
It is necessary to update the translation files into Ukrainian in the project.  I dropped them above the 11th.

jannikstdl commented:
@tjbck looks good, i would test it if i find time but if @asedmammad says it's fine, merge it.

MorrisLu-Taipei commented:
Hi all, thanks for your efforts, I also completed my translation file for Traditional Chinese (Taiwan) as link below
https://drive.google.com/file/d/17CF5WDtwk_V_SRJ4kPfcf8KYMK6v3I7S/view?usp=drive_link

Please also let me know if multi language support ready. thanks
Translation by Morris Lu, National Taiwan University of Science and Technology

jannikstdl commented:
@MorrisLu-Taipei Thanks, added. We can for now only import 2 char language codes so i imported it as "zh" not "zh-tw" for Taiwan.

kawais commented:
Hi, This is my translation file for Simplified Chinese, I hope it is useful
[https://drive.google.com/file/d/19vwGtzjWMcMSIX5_eDiVwaH8XoQHnl1u/view?usp=sharing](https://drive.google.com/file/d/19vwGtzjWMcMSIX5_eDiVwaH8XoQHnl1u/view?usp=sharing)

SimonOriginal commented:
I've updated the Ukrainian translation file - "https://drive.google.com/drive/u/0/folders/17Yx3d0b4JqHr9va7kvHkQDzhbcrHQS3V"

jannikstdl commented:
@kawais @MorrisLu-Taipei Added both now, simplified and traditional language codes:
![image](https://github.com/open-webui/open-webui/assets/69747628/5ab11ed9-524d-4d96-ace4-70e5b0b0e2e2)

@asedmammad To support multiple variants of one language like below i migrated to `xx-XX` lang codes.
Changed CONTRIBUTION.md also

@SimonOriginal updated your new translation keys too

SimonOriginal commented:
Thank you @jannikstdl. I would like to draw attention to a noticeable problem in the design related to the uneven feel of the font. We need to figure out why it looks smooth and pleasant in some places, and in other places it looks sharp and doesn't match the overall aesthetic impression (I noticed it only in the Ukrainian localisation).

Besides, there are still a few places that need to be added to support the localisation. I have added the keys of these places to the Ukrainian translation beforehand (I don't want to bother you too much). Thank you for your work.

justinh-rahb commented:
> Besides, there are still a few places that need to be added to support the localisation. I have added the keys of these places to the Ukrainian translation beforehand (I don't want to bother you too much). Thank you for your work.

Cyrillic characters can be difficult, we'd need to ensure the font(s) used properly support them.

jannikstdl commented:
> Thank you @jannikstdl. I would like to draw attention to a noticeable problem in the design related to the uneven feel of the font. We need to figure out why it looks smooth and pleasant in some places, and in other places it looks sharp and doesn't match the overall aesthetic impression (I noticed it only in the Ukrainian localisation).

@SimonOriginal yes i saw your screenshots, don't know the cause yet.
I will look deeper, maybe i find a way to fix this.


> Besides, there are still a few places that need to be added to support the localisation. I have added the keys of these places to the Ukrainian translation beforehand (I don't want to bother you too much). Thank you for your work.

What keys do you exactly mean?

SimonOriginal commented:
Here are the screenshots: 
![зображення](https://github.com/open-webui/open-webui/assets/94782611/e5711901-f888-495d-b4af-233f2623e4f7)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/c41364e3-7281-48dc-bd2b-9e914bf46ee8)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/a77ef9c8-8b7e-41f4-9c69-70335ab787c0)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/8d5bf54d-a8d6-4f98-8b62-840428131ae4)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/6ef8150a-7f9b-445a-bffc-e716371535be)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/4b333273-6f43-46d4-85c3-21d80ab05eca)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/2944f435-c222-490d-a50c-427e37c25539)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/cb766e09-dea7-4186-a75f-30fe674ebedf)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/c1ff8b92-a1d9-441c-98dd-4f7283b9a6a6)
![зображення](https://github.com/open-webui/open-webui/assets/94782611/e5a54a91-879d-4f49-a241-1cadac433eeb)
Sorry if I'm being too picky.

jannikstdl commented:
All good we appreciate your effort. I will change that.

MorrisLu-Taipei commented:
Hi all, sorry, I am a newbie in this localization
how can i enable this multi-language support in my env or setting? 
Or, I have to wait for nex t version to support this new feature.
thanks

jannikstdl commented:
Added missing translation keys (some are related to the newest release)
![image](https://github.com/open-webui/open-webui/assets/69747628/2d27030a-6836-4b6f-aca8-1ed6be45dc54)

@SimonOriginal The message "The requested action.." can not be changed for now because this comes from the backend.

justinh-rahb commented:
@jannikstdl maybe we could get a preview docker build pushed somewhere for folks to test this out? I'll see about getting it into my repo.

SimonOriginal commented:
I built the project. Everything is super! Just need to replace "Open WebUI Version": "Open WebUI Version", replace it with "Version": "Version" to apply the translation. 
![зображення](https://github.com/open-webui/open-webui/assets/94782611/16fa180b-a5f0-4cf6-aa4c-d139447df00a)

justinh-rahb commented:
I've pushed @jannikstdl's most recent commit to my own repo and merged it to my `dev` branch, here's the Docker image URL for anyone else that would like to test this out now:
```
docker pull ghcr.io/justinh-rahb/open-webui:dev
```

jannikstdl commented:
It's ready to be tested, thanks for the work everyone!!
> ```
> docker pull ghcr.io/justinh-rahb/open-webui:dev
> ```
 After some user testing this should be ready to merge @tjbck 
@asedmammad also marked it as ready for review.

asedmammad commented:
@jannikstdl 🌟 Thanks for addressing the issues and updating the PR.
Regrettably, I was unable to respond promptly due to an unexpectedly heavy workload at the office.
The changes LGTM.

SimonOriginal commented:
<p>Here are a couple more that may have been overlooked.</p>
<img src="https://github.com/open-webui/open-webui/assets/94782611/7bfdad67-50f3-40bb-9c41-c94245caa3f9" alt="зображення" width="200">
<img src="https://github.com/open-webui/open-webui/assets/94782611/be392934-2994-4f3a-9983-c495afa3cbc7" alt="зображення" width="200">
<img src="https://github.com/open-webui/open-webui/assets/94782611/349113c1-f2dc-4402-ad36-821beddbd1a1" alt="зображення" width="200">
<p>I hope that's it.</p>

jannikstdl commented:
> <img alt="зображення" width="200" src="https://private-user-images.githubusercontent.com/94782611/312922664-349113c1-f2dc-4402-ad36-821beddbd1a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0NDQxMDksIm5iZiI6MTcxMDQ0MzgwOSwicGF0aCI6Ii85NDc4MjYxMS8zMTI5MjI2NjQtMzQ5MTEzYzEtZjJkYy00NDAyLWFkMzYtODIxYmVkZGJkMWExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE0VDE5MTY0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkZmI0ZTgyYjZkMjY5NWJkNTE1MDQxYzAwNWUwMmNlM2M5ZjNlN2YxZmMyMWFlODIyZDU4MDBhMGU1YzA0OGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.h6kJ_mdhmHvp5f3RWPzb9pPOTA0A01msKddREfUTWW4">
Good catch!
Do you mean the "latest" tag?

SimonOriginal commented:
No, you only need to replace one line there and localisation will work. 
<img src="https://github.com/open-webui/open-webui/assets/94782611/7943fd59-3424-459f-9dc5-b249973245a4" alt="зображення" width="500">
Delete this line. 
<img src="https://github.com/open-webui/open-webui/assets/94782611/383ba36d-e32c-4b50-8567-d3dd7a89ffdd" width="500">
And replace it with this one - "Version": "Version".

SimonOriginal commented:
Here's another translation for the newly added keys.

"Model(s) Whitelisted": "Модель(і) білого списку",
"(latest)": "(остання)",

justinh-rahb commented:
New pushed to my repo, should be about ~20 minutes til Docker image is ready.

SimonOriginal commented:
"Settings saved successfully!": "Налаштування успішно збережено!",

jannikstdl commented:
@asedmammad The parser is removing the 

`"Enter a user message here": "Enter a user message here",`
and 
`"Enter an assistant message here": "Enter an assistant message here",`

keys because it thinks this are unused keys since it's implemented like that:

![image](https://github.com/open-webui/open-webui/assets/69747628/17a0c77e-5810-4292-b287-f1baafdce087)

asedmammad commented:
> @asedmammad The parser is removing the
> 
> `"Enter a user message here": "Enter a user message here",` and `"Enter an assistant message here": "Enter an assistant message here",`

@jannikstdl  I think we should change that to a more parser friendly format(!), something like this:
```
`Enter {{role}} message here`, {role: message.role === 'user' ? 'a user' : 'an assistant'}
```
Will update and bring back the keys.

asedmammad commented:
> @jannikstdl I think we should change that to a more parser friendly format(!), something like this:
> ```
> `Enter {{role}} message here`, {role: message.role === 'user' ? $i18n.t('a user') : $i18n.t('an assistant')
> ```

`i18next-parser` does not seem to detect `a user` and `an assistant` keys when they are on the same line (prettier puts them on the same line),  a simple work around (from `i18next-parser` docs) could be adding these above ternary expressions:
```
<!-- $i18n.t('a user') -->
<!-- $i18n.t('an assistant') -->
<textarea
	id="{message.role}-{idx}-textarea"
	bind:this={textAreaElement}
	class="w-full bg-transparent outline-none rounded-lg p-2 text-sm resize-none overflow-hidden"
	placeholder={$i18n.t(`Enter {{role}} message here`, {
		role: message.role === 'user' ? $i18n.t('a user') : $i18n.t('an assistant')
	})}
```
Another option is to change `i18next-parser` config to not remove obsolete keys.
I'm looking for other options (currently investigating `i18next-parser` code), @jannikstdl  Do you have any other idea?

jannikstdl commented:
That workaround looks good to me. 

I initially also was thinking about setting the parser config to not remove objects but I guess that is a feature we should let on.  If someone in future removes pages or components it is cleaner to remove unused keys.

SimonOriginal commented:
"Enter {{role}} message here": "Введіть повідомлення {{role}} тут",
"a user": "користувача",
"an assistant": "асистента",

justinh-rahb commented:
New dev docker container in my repo.

tjbck commented:
Just tried it out, and this is beyond amazing! Thanks for all the effort and being a part of this journey guys! We're one step closer to truly democratising access to AI for all! I'll merge this to dev branch right now and feel free to make additional PRs in case anything needs to be fixed!
